tests/rptest/tests/internal_topic_protection_test.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2022 Redpanda Data, Inc.
+#
+# Use of this software is governed by the Business Source License
+# included in the file licenses/BSL.md
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0
+
+import subprocess
+import time
+
+from rptest.services.cluster import cluster
+from rptest.tests.redpanda_test import RedpandaTest
+from rptest.clients.rpk import RpkTool, RpkException
+from rptest.clients.kafka_cli_tools import KafkaCliTools
+from rptest.clients.kafka_cat import KafkaCat
+from ducktape.utils.util import wait_until
+
+
+class InternalTopicProtectionTest(RedpandaTest):
+    """
+    Verify that the `kafka_nodelete_topics` and `kafka_noproduce_topics`
+    configuration properties function as intended.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, extra_rp_conf={}, **kwargs)
+
+        self.kafka_tools = KafkaCliTools(self.redpanda)
+        self.kafka_cat = KafkaCat(self.redpanda)
+        self.rpk = RpkTool(self.redpanda)
+
+    @cluster(num_nodes=3)
+    def kafka_noproduce_topics_test(self):
+        def test_with_produce_fn(produce_fn, failure_exception_type):
+            test_topic = "noproduce_topic"
+            self.kafka_tools.create_topic_with_config(test_topic, 3, 3, {})
+
+            wait_until(lambda: test_topic in self.rpk.list_topics(),
+                       timeout_sec=90,
+                       backoff_sec=3)
+
+            # Ensure topic can't be produced to via the Kafka API when
+            # it's in the kafka_noproduce_topics list.
+            self.rpk.cluster_config_set("kafka_noproduce_topics",
+                                        str([test_topic]))
+
+            try:
+                produce_fn(test_topic, "test_msg")
+            except Exception as e:
+                assert type(e) == failure_exception_type
+            else:
+                assert False, "Call to delete topic returned sucess"
+
+            try:
+                self.rpk.consume(test_topic, n=1, timeout=30)
```

Comment:
Is this going to wait 30 seconds before proceeding?

Maybe better to read the hwm of the partition as a way to immediately check for no messages, rather than having a consumer time out.

Reply:
Would the HWM mark be updated right after a produce? I assume rpk queries the partition leader for the HWM so it should be safe to check it before and after a produce to the partition to validate that the produce failed?

Reply:
Right, exactly that.

Reply:
Great, switched to using the hwm.